src/plan_expand_hypertable.c

```diff
@@ -80,100 +80,95 @@ is_chunk_exclusion_func(Node *node)
  * to be as comprehensive as the PG native derivation. This is inspired
  * by the derivation in `deconstruct_recurse` in PG
  *
- * We also try to detect explicit chunk exclusion and validate it is properly used. If
- * explicit chunk exclusion function is detected we'll skip adding restrictions since
- * they will not be used anyway.
+ * We also try to detect explicit chunk exclusion and validate it is properly used.
  *
- * This method is modifying the query tree by removing `chunks_in` function node
- * from the WHERE clause. This is done to prevent the function from making the
- * planner believe that an index-only scan is not possible (since function is
- * only used to exclude chunks we actually don't need that function to run so it
- * can be removed).
+ * This function removes chunks_in from the list of quals, because chunks_in is
+ * just used as marker function to trigger explicit chunk exclusion and the function
+ * will throw an error when executed.
  */
 static Node *
-collect_quals_mutator(Node *node, CollectQualCtx *ctx)
+process_quals(Node *quals, CollectQualCtx *ctx)
 {
-	if (node == NULL)
-		return NULL;
+	ListCell *lc;
+	ListCell *prev = NULL;
+	ListCell *next = NULL;
 
-	if (IsA(node, FromExpr))
+	for (lc = list_head((List *) quals); lc != NULL; prev = lc, lc = next)
 	{
-		FromExpr *f = (FromExpr *) node;
-		ListCell *lc;
-		ListCell *prev = NULL;
-		ListCell *next = NULL;
-		bool func_removed = false;
+		Node *qual = (Node *) lfirst(lc);
+		RestrictInfo *restrictinfo;
+		Relids relids = pull_varnos(qual);
+
+		/*
+		 * keep track of next item separately because we modify
+		 * list in place when removing chunks_in calls
+		 */
+		next = lnext(lc);
+
+		/*
+		 * skip expressions not for current rel
+		 */
+		if (bms_num_members(relids) != 1 || !bms_is_member(ctx->rel->relid, relids))
+			continue;
 
-		for (lc = list_head((List *) f->quals); lc != NULL;)
+		if (is_chunk_exclusion_func(qual))
+		{
+			FuncExpr *func_expr = (FuncExpr *) qual;
+
+			/* validation */
+			if (ctx->chunk_exclusion_func != NULL)
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("only one chunks_in call is allowed per hypertable")));
+
+			Assert(func_expr->args->length == 2);
+			if (!IsA(linitial(func_expr->args), Var))
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("first parameter for chunks_in function needs to be a record")));
+
+			ctx->chunk_exclusion_func = func_expr;
+			quals = (Node *) list_delete_cell((List *) quals, lc, prev);
+			lc = prev;
+		}
+		else
```

Comment:
should this be `else if(ctx->chunk_exclusion_func == NULL)`

Reply:
I've changed this because this leads to inconsistent results. Now it will always build restrictinfo for all quals that qualify, previously it would stop building restrictinfo depending on the position of the chunks_in call in the quals